feat(forms): add email validator

DzmitryShylovich opened the pull request:
Closes #13706

Regexp from ng1 [link](https://github.com/angular/angular.js/blob/a24777a2c4ad2ac087d9e3aa278fa2e61e8cc740/src/ng/directive/input.js#L27).

Commit:
feat(forms): add email validator

Closes #13706

vicb commented on modules/@angular/forms/src/directives/validators.ts:
```diff
@@ -107,6 +107,51 @@ export class CheckboxRequiredValidator extends RequiredValidator {
+  host: {'[attr.email]': 'email ? email : null'}
+})
+export class EmailValidator implements Validator {
+  private _email: boolean;
```
`_email` sounds wrong for a boolean value. Why does the getter return `string|boolean` ?

DzmitryShylovich replied:
> _email sounds wrong for a boolean value.

agreed :)

> Why does the getter return string|boolean ?

because `set` and `get` accessor must have the same type

vicb commented on modules/@angular/forms/src/directives/validators.ts:
```diff
@@ -107,6 +107,51 @@ export class CheckboxRequiredValidator extends RequiredValidator {
+  get email(): boolean|string { return this._email; }
+
+  set email(value: boolean|string) {
+    this._email = value != null && value !== false && `${value}` !== 'false';
```
`\`${value}\` !== 'false'` why not a simple string test at this point ?

DzmitryShylovich replied:
I made it similar to [RequiredValidator](https://github.com/angular/angular/blob/master/modules/%40angular/forms/src/directives/validators.ts#L74)

vicb commented on modules/@angular/forms/src/directives/validators.ts:
````diff
@@ -107,6 +107,51 @@ export class CheckboxRequiredValidator extends RequiredValidator {
+ * ### Example
+ *
+ * ```
+ * <input type="email" name="active" ngModel email>
````
document the prop value ?

vicb commented on modules/@angular/forms/src/directives/validators.ts:
```diff
@@ -107,6 +107,51 @@ export class CheckboxRequiredValidator extends RequiredValidator {
+  host: {'[attr.email]': 'email ? email : null'}
+})
+export class EmailValidator implements Validator {
+  private _validate: boolean;
```
`_active` ?

vicb commented on modules/@angular/forms/src/directives/validators.ts:
```diff
@@ -107,6 +107,51 @@ export class CheckboxRequiredValidator extends RequiredValidator {
+  get email(): boolean|string { return this._validate; }
+
+  set email(value: boolean|string) {
+    this._validate = value != null && value !== false && value !== 'false';
```
probably this should be
```TS
this._disabled = value == null || value === false || value === 'false';
```

and I'm not actually clear if null is good here ?

DzmitryShylovich replied:
I think it's better to check for truthy values
```
this._enabled = value === '' || value === true || value === 'true';
```

vicb replied:
depends what you mean:
- truthy means `!!x === true` so `'false'` is truthy,
- your code would be inconsistent with other validator.

Here I think what we want to validate when
- `<input email>` -> `''`
- `<input [email]=[a value]>` -> value should be truthy

so to me this should be `enabled = !!v || v === ''` but again inconsistent.

@kara WDYT ?

vicb commented on modules/@angular/forms/src/directives/validators.ts:
```diff
@@ -107,6 +107,51 @@ export class CheckboxRequiredValidator extends RequiredValidator {
+@Directive({
+  selector: '[email][formControlName],[email][formControl],[email][ngModel]',
+  providers: [EMAIL_VALIDATOR],
+  host: {'[attr.email]': 'email ? email : null'}
```
is there such thing as an `email` attribute in the HTML spec ?

DzmitryShylovich replied:
nope :)

DzmitryShylovich commented:
@vicb comments are addressed. Also added a test.

> document the prop value ?

I added usage example. It has the same values as other directives so it shouldn't be confusing.

yoricktran commented:
Hey there - the validation regex seems to be incorrect. Mine validates abc@web as correct although domain is missing.

yechenchao commented:
I think the regex is correct as they just follow the standard email format, which include email under local domain, but not accepted in most website or apps.

Toxicable commented:
It is correct, see https://github.com/angular/angular/pull/18658#issuecomment-321923635
